File: cicd_dagger_contract/golang/deploy.example.go
package main

import (
	"context"
	"encoding/json"
	"time"

	"dagger/goserv/internal/dagger"

	"dagger.io/dagger/dag"
)

// Deploy installs the Helm chart from a Helm repository to a Kubernetes cluster
// +cache = "never"
func (m *Goserv) Deploy(
	ctx context.Context,
	// Source directory containing the project
	source *dagger.Directory,
	// +optional
	// Kubernetes config file content
	kubeconfig *dagger.Secret,
	// +optional
	// AWS configuration file content
	awsconfig *dagger.Secret,
	// +optional
	// Helm chart repository URL (default: oci://ttl.sh)
	helmRepository string,
	// +optional
	// Container repository URL (default: ttl.sh)
	containerRepository string,
	// +optional
	// Build as release candidate (appends -rc to version tag)
	releaseCandidate bool,
) (*dagger.File, error) {
	// Perform deployment (helm install/upgrade)
	// ... deployment logic here ...

	// Create deployment context
	deploymentContext := map[string]interface{}{
		"timestamp":      time.Now().Format(time.RFC3339),
		"endpoint":       "http://goserv.default.svc.cluster.local:8080",
		"releaseName":    "goserv",
		"namespace":      "default",
		"chartVersion":   "0.1.0",
		"imageReference": imageRef,
	}

	contextJSON, err := json.MarshalIndent(deploymentContext, "", "  ")
	if err != nil {
		return nil, err
	}

	// Return as file
	return dag.Directory().
		WithNewFile("deployment-context.json", string(contextJSON)).
		File("deployment-context.json"), nil
}

File: cicd_dagger_contract/java/Deploy.example.java
package cicd;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.dagger.client.Container;
import io.dagger.client.Dagger;
import io.dagger.client.Directory;
import io.dagger.client.File;
import io.dagger.client.Secret;
import java.time.Instant;

public class Deploy {

    /**
     * Deploy installs the Helm chart from a Helm repository to a Kubernetes cluster
     * Note: This function should have cache = "never" configuration
     *
     * @param source Source directory containing the project
     * @param kubeconfig Kubernetes config file content
     * @param awsconfig AWS configuration file content
     * @param helmRepository Helm chart repository URL (default: oci://ttl.sh)
     * @param containerRepository Container repository URL (default: ttl.sh)
     * @param releaseCandidate Build as release candidate (appends -rc to version tag)
     * @return File containing deployment context
     */
    public File deploy(
            Directory source,
            Secret kubeconfig,
            Secret awsconfig,
            String helmRepository,
            String containerRepository,
            boolean releaseCandidate) throws Exception {
        // Perform deployment (helm install/upgrade)
        // ... deployment logic here ...

        // Create deployment context
        ObjectMapper mapper = new ObjectMapper();
        ObjectNode deploymentContext = mapper.createObjectNode();
        deploymentContext.put("timestamp", Instant.now().toString());
        deploymentContext.put("endpoint", "http://goserv.default.svc.cluster.local:8080");
        deploymentContext.put("releaseName", "goserv");
        deploymentContext.put("namespace", "default");
        deploymentContext.put("chartVersion", "0.1.0");
        deploymentContext.put("imageReference", imageRef);

        String contextJson = mapper.writerWithDefaultPrettyPrinter()
            .writeValueAsString(deploymentContext);

        // Return as file
        return Dagger.dag().directory()
            .withNewFile("deployment-context.json", contextJson)
            .file("deployment-context.json");
    }
}
